contracts/pallet-ibc/src/ics20/mod.rs

```diff
@@ -561,11 +561,187 @@ pub fn full_ibc_denom(packet: &Packet, mut token: PrefixedCoin) -> String {
 use ibc::applications::transfer::error::Error as Ics20Error;
 
 pub trait HandleMemo<T: Config> {
-	fn execute_memo(packet_data: &PacketData, receiver: T::AccountId) -> Result<(), Ics20Error>;
+	fn execute_memo(
+		packet: &Packet,
+		packet_data: &PacketData,
+		receiver: T::AccountId,
+	) -> Result<(), Ics20Error>;
 }
 
 impl<T: Config> HandleMemo<T> for () {
-	fn execute_memo(_packet_data: &PacketData, _receiver: T::AccountId) -> Result<(), Ics20Error> {
+	fn execute_memo(
+		_packet: &Packet,
+		_packet_data: &PacketData,
+		_receiver: T::AccountId,
+	) -> Result<(), Ics20Error> {
 		Ok(())
 	}
 }
+
+use frame_system::RawOrigin;
+use scale_info::prelude::boxed::Box;
+#[derive(serde::Serialize, serde::Deserialize, Debug)]
+struct MemoForward {
+	receiver: String,
+	port: String,
+	channel: String,
+	timeout: String,
+	retries: u64,
+	next: Option<Box<MemoForward>>,
```

Comment:
would it make sense to have some validation logic to ensure that this isn't a graph that contains a loop?

Reply:
I think it's not necessary at the moment. It won't lead to an infinite loop and also it's unlikely to happen if we have validation on front-end side

Reply:
this could be a potential "vulnerability". Not critical in terms of funds being lost, but unnecessary usage of resources, right?

Reply:
True.

Reply:
@blasrodri we always extract `.next` from the memo, so the loop is not possible. Please confirm @RustNinja

Reply:
we extract next memo to pass it further but still the loop is possible if someone sign tx on cosmos and specify the memo with a loop to move assets back to cosmos. so now there is no validation to stop memo execution if the loop inside the memo. @blasrodri

Reply:
But once the transaction is returned, it won't contain any memo, right? Because we specified only one hop (cosmos->X->cosmos) and I don't see any reason to not allow it if it's not forbidden in the standard

Reply:
Let me know what to do with it. As i know cosmos chains does not reject any circuit memo.